### ISSUE TITLE
Cannot select Generic tests from command line

### ISSUE THREAD

**Issue by x2764tech** [nunit/nunit#1049]
When running `NUnit Console Runner 3.0.5797` with a test selection that includes a generic type argument, I get and error an no tests are run:

    >nunit3-console.exe  WebTests.dll --where "test == WebTests._3DSecure.Test_3DSecure<BrowserStackDriverBuilder>.Test_MC_CardholderEnrolled_AuthenticationNotAbleToComplete"
    NUnit Console Runner 3.0.5797
    Copyright (C) 2015 Charlie Poole

    Runtime Environment
       OS Version: Microsoft Windows NT 6.3.9600.0
      CLR Version: 4.0.30319.42000

    Test Files
        WebTests.dll

    Test Filters
        Where: test == WebTests._3DSecure.Test_3DSecure<BrowserStackDriverBuilder>.Test_MC_CardholderEnrolled_AuthenticationNotAbleToComplete

    System.Xml.XmlException: The 'BrowserStackDriverBuilder' start tag on line 1 position 48 does not match the end tag of 'test'. Line 1, position 135.
       at System.Xml.XmlTextReaderImpl.Throw(Exception e)
       at System.Xml.XmlTextReaderImpl.ThrowTagMismatch(NodeData startTag)
       at System.Xml.XmlTextReaderImpl.ParseEndElement()
       at System.Xml.XmlTextReaderImpl.ParseElementContent()
       at System.Xml.XmlLoader.LoadNode(Boolean skipOverWhitespace)
       at System.Xml.XmlLoader.LoadDocSequence(XmlDocument parentDoc)
       at System.Xml.XmlDocument.Load(XmlReader reader)
       at System.Xml.XmlDocument.LoadXml(String xml)
       at NUnit.Engine.Internal.XmlHelper.CreateXmlNode(String xml)
       at NUnit.Engine.Runners.MasterTestRunner.InsertFilterElement(XmlNode resultNode, TestFilter filter)
       at NUnit.Engine.Runners.MasterTestRunner.RunTests(ITestEventListener listener, TestFilter filter)
       at NUnit.Engine.Runners.MasterTestRunner.NUnit.Engine.ITestRunner.Run(ITestEventListener listener, TestFilter filter)
       at NUnit.ConsoleRunner.ConsoleRunner.RunTests(TestPackage package, TestFilter filter)
       at NUnit.ConsoleRunner.Program.Main(String[] args)

If I try xml/html escaping the test name, I get:
```
NUnit.Engine.TestSelectionParserException: Unexpected token '&' at position 0 in selection expression.
   at NUnit.Engine.TestSelectionParser.Expect(Token[] valid)
   at NUnit.Engine.TestSelectionParser.ParseFilterElement()
   at NUnit.Engine.TestSelectionParser.ParseFilterTerm()
   at NUnit.Engine.TestSelectionParser.ParseFilterExpression()
   at NUnit.Engine.TestFilterBuilder.GetFilter()
   at NUnit.ConsoleRunner.ConsoleRunner.Execute()
   at NUnit.ConsoleRunner.Program.Main(String[] args)
```
Specifying the name name of the test using the deprecated `--test` argument results in an error too:

	C:\>nunit3-console.exe WebTests.dll --test "WebTests._3DSecure.Test_3DSecure<BrowserStackDriverBuilder>.Test_MC_CardholderEnrolled_AuthenticationNotAbleToComplete"
	NUnit Console Runner 3.0.5797
	Copyright (C) 2015 Charlie Poole

	Runtime Environment
	   OS Version: Microsoft Windows NT 6.3.9600.0
	  CLR Version: 4.0.30319.42000

	Test Files
		WebTests.dll

	Test Filters
		Test: WebTests._3DSecure.Test_3DSecure<BrowserStackDriverBuilder>.Test_MC_CardholderEnrolled_AuthenticationNotAbleTo
	Complete

	System.Xml.XmlException: The 'BrowserStackDriverBuilder' start tag on line 1 position 48 does not match the end tag of '
	test'. Line 1, position 135.
	   at System.Xml.XmlTextReaderImpl.Throw(Exception e)
	   at System.Xml.XmlTextReaderImpl.ThrowTagMismatch(NodeData startTag)
	   at System.Xml.XmlTextReaderImpl.ParseEndElement()
	   at System.Xml.XmlTextReaderImpl.ParseElementContent()
	   at System.Xml.XmlLoader.LoadNode(Boolean skipOverWhitespace)
	   at System.Xml.XmlLoader.LoadDocSequence(XmlDocument parentDoc)
	   at System.Xml.XmlDocument.Load(XmlReader reader)
	   at System.Xml.XmlDocument.LoadXml(String xml)
	   at NUnit.Engine.Internal.XmlHelper.CreateXmlNode(String xml)
	   at NUnit.Engine.Runners.MasterTestRunner.InsertFilterElement(XmlNode resultNode, TestFilter filter)
	   at NUnit.Engine.Runners.MasterTestRunner.RunTests(ITestEventListener listener, TestFilter filter)
	   at NUnit.Engine.Runners.MasterTestRunner.NUnit.Engine.ITestRunner.Run(ITestEventListener listener, TestFilter filter)
	   at NUnit.ConsoleRunner.ConsoleRunner.RunTests(TestPackage package, TestFilter filter)
	   at NUnit.ConsoleRunner.Program.Main(String[] args)

I can't find any test cases related to parsing generic type arguments, so I don't know if this is expected behaviour.

**Comment by CharliePoole** [comment]
I was able to run a generic fixture in the NUnit test suite using this command:

```
nunit3-console net-4.5\nunit.framework.tests.dll --where "test==/NUnit.Framework.Internal.GenericTestFixture_Numeric\&lt;T\&gt;/"
```

The double quotes enclose the argument and are dropped before NUnit gets the contents. The right hand side of the expression is a quoted string, using / as the delimiter because the command line would require escaping quotes. The backslashes allow ampersand to appear in the command line.

This indicates that what you want to do is possible with the existing TSL but difficult because command-line escaping and xml escaping have to be combined. It may be easier to use the class, method or name keywords.

**Comment by CharliePoole** [comment]
After some thought, I'm re-opening this. The necessary command-line escaping is required by the OS in order to pass the proper string to NUnit, so we can't do anything about that. However, it really should not be necessary for the user to do the XML escaping. NUnit should do it.

**Comment by x2764tech** [comment]
I was wondering if the explore option could show the escaped name, but I guess this would be different depending on the command shell (bash/PowerShell/windows command line, etc), so it might not really be feasible/sensible for the runner to try to output escaped names.

Maybe even a documentation note on the Test Selection Language page might be enough?

**Comment by CharliePoole** [comment]
If you mean the xml elements, I don't think so. The use of xml for filtering is an internal implementation detail.

As you say, command-line escaping depends on the OS. A note in the docs is a good idea.

**Comment by rprouse** [comment]
Closed with PR #1084

**Comment by x2764tech** [comment]
Thanks for sorting this out
